packages/contract/src/pay/DaimoPayRelayer.sol

```diff
@@ -2,83 +2,122 @@
 pragma solidity ^0.8.12;
 
 import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
-import "openzeppelin-contracts/contracts/access/Ownable2Step.sol";
 import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
 
 import "./DaimoPay.sol";
 import "./TokenUtils.sol";
 
 /*
  * Relayer contract that funds completes DaimoPay intents.
  */
-contract DaimoPayRelayer is Ownable2Step {
+contract DaimoPayRelayer is AccessControl {
     using SafeERC20 for IERC20;
 
-    constructor(address _owner) Ownable(_owner) {}
+    bytes32 public constant RELAYER_ROLE = keccak256("RELAYER_ROLE");
+
+    event SwapAndTip(
+        address indexed requiredTokenIn,
+        uint256 suppliedAmountIn,
+        address indexed requiredTokenOut,
+        uint256 swapAmountOut,
+        uint256 maxPreTip,
+        uint256 maxPostTip
+    );
+
+    constructor(address admin) {
+        _grantRole(DEFAULT_ADMIN_ROLE, admin);
+        _grantRole(RELAYER_ROLE, admin);
+    }
+
+    // Add a new address that can use the relayer functions.
+    function grantRelayerRole(
+        address relayer
+    ) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        _grantRole(RELAYER_ROLE, relayer);
+    }
+
+    // Withdraws an amount of tokens from the contract to the admin.
+    function withdrawAmount(
+        IERC20 token,
+        uint256 amount
+    ) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        TokenUtils.transfer(token, payable(msg.sender), amount);
+    }
+
+    // Withdraws the full balance of a token from the contract to the admin.
+    function withdrawBalance(IERC20 token) public onlyRole(DEFAULT_ADMIN_ROLE) {
+        TokenUtils.transferBalance(token, payable(msg.sender));
```

Comment:
maybe have this return `uint256`, then `return TokenUtils.transferBalance...`

this way if admin needs to send the token balance on to anywhere else they know the amount cleanly